### PR TITLE
MPS Support

### PR DESCRIPTION
**Pull request by joecummings**
Copy of #790 but with proper rebase

@maximegmd is the real hero here.

### CODE REVIEW

**Comment by joecummings** [comment]
Update hitting a blocker - can't get this running on M1 Mac with 32GB RAM:

```
(joe-torchtune) jrcummings@jrcummings-mbp joe-torchtune % tune run full_finetune_single_device --config gemma/2B_full_single_device device=mps optimizer._component_=torch.optim.SGD
W0726 16:52:51.106000 8571538432 torch/distributed/elastic/multiprocessing/redirects.py:28] NOTE: Redirects are currently not supported in Windows or MacOs.
INFO:torchtune.utils.logging:Running FullFinetuneRecipeSingleDevice with resolved config:

batch_size: 1
checkpointer:
  _component_: torchtune.utils.FullModelHFCheckpointer
  checkpoint_dir: /tmp/gemma-2b/
  checkpoint_files:
  - model-00001-of-00002.safetensors
  - model-00002-of-00002.safetensors
  model_type: GEMMA
  output_dir: /tmp/gemma-2b
  recipe_checkpoint: null
compile: false
dataset:
  _component_: torchtune.datasets.alpaca_dataset
  train_on_input: true
device: mps
dtype: bf16
enable_activation_checkpointing: true
epochs: 1
gradient_accumulation_steps: 1
log_every_n_steps: 1
log_peak_memory_stats: true
loss:
  _component_: torch.nn.CrossEntropyLoss
max_steps_per_epoch: null
metric_logger:
  _component_: torchtune.utils.metric_logging.DiskLogger
  log_dir: /tmp/gemma-2b/logs
model:
  _component_: torchtune.models.gemma.gemma_2b
optimizer:
  _component_: torch.optim.SGD
  lr: 2.0e-05
optimizer_in_bwd: false
output_dir: /tmp/gemma-2b/logs
resume_from_checkpoint: false
seed: null
shuffle: true
tokenizer:
  _component_: torchtune.models.gemma.gemma_tokenizer
  path: /tmp/gemma-2b/tokenizer.model

DEBUG:torchtune.utils.logging:Setting manual seed to local seed 2219507826. Local seed is seed + rank = 2219507826 + 0
Writing logs to /tmp/gemma-2b/logs/log_1722027171.txt
INFO:torchtune.utils.logging:Model is initialized with precision torch.bfloat16.
INFO:torchtune.utils.logging:Tokenizer is initialized from file.
INFO:torchtune.utils.logging:Optimizer is initialized.
INFO:torchtune.utils.logging:Loss is initialized.
INFO:torchtune.utils.logging:Dataset and Sampler are initialized.
  0%|                                                                                                                                                                                                                      | 0/52002 [00:00<?, ?it/s]/Users/jrcummings/miniconda3/envs/joe-torchtune/lib/python3.11/site-packages/torch/utils/checkpoint.py:1399: FutureWarning: `torch.cpu.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cpu', args...)` instead.
  with device_autocast_ctx, torch.cpu.amp.autocast(**cpu_autocast_kwargs), recompute_context:  # type: ignore[attr-defined]
1|1|Loss: 2.1554598808288574:   0%|                                                                                                                                                                             | 1/52002 [00:03<52:31:28,  3.64s/it]/AppleInternal/Library/BuildRoots/91a344b1-f985-11ee-b563-fe8bc7981bff/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:869: failed assertion `[MPSNDArray, initWithBuffer:descriptor:] Error: buffer is not large enough. Must be 393216 bytes
'
zsh: abort      tune run full_finetune_single_device --config gemma/2B_full_single_device
(joe-torchtune) jrcummings@jrcummings-mbp joe-torchtune % /Users/jrcummings/miniconda3/envs/joe-torchtune/lib/python3.11/multiprocessing/resource_tracker.py:254: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

Looks like it could be related to https://github.com/pytorch/pytorch/issues/87351? But that truly seems unlikely as this issue is from October 2022.

**Comment by joecummings** [comment]
With PyTorch nightlies, I manage to run 4 steps before loss NaNs and buffer issue comes back:

```
1|4|Loss: nan:   0%|                                                                                                                                                                                            | 4/52002 [00:26<95:05:25,  6.58s/it]/AppleInternal/Library/BuildRoots/91a344b1-f985-11ee-b563-fe8bc7981bff/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:869: failed assertion `[MPSNDArray, initWithBuffer:descriptor:] Error: buffer is not large enough. Must be 163840 bytes
'
zsh: abort      tune run full_finetune_single_device --config gemma/2B_full_single_device
(joe-torchtune) jrcummings@jrcummings-mbp joe-torchtune % /Users/jrcummings/miniconda3/envs/joe-torchtune/lib/python3.11/multiprocessing/resource_tracker.py:254: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

**Comment by maximegmd** [comment]
That's odd, I was able to train for hours with no issue. I can try a run tomorrow (64GB maybe this is why?)

**Comment by byjlw** [comment]
I can test it on my machine but i need details on which commands to run (general instructions). 
We need this in general for this readme. 
Either the main readme.md or a mps.md file in the cookbooks directory. 
Once the PR has been updated with this i'll run through it on my M1 Max 64GB

**Comment by SalmanMohammadi** [comment]
- [ ] Default to `aot_eager` for compile on mps.
- [ ] Ensure PyTorch nightly version/check for version (at least up-to-date with fix for https://github.com/pytorch/pytorch/issues/130613#issuecomment-2226549088)
- [ ] Document install instructions:
    - [ ] torchao needs to be installed from source
    - [ ] Dev install needs to exclude bitsandbytes
    - [ ] All installs need torch nightly
- [ ] Add memory tracking in `get_memory_stats`
- [ ] `test_nf4_linear` crashes out due to importing bitsandbytes. Need to try catch? And mark the tests to skip if mps.

@joecummings things I usually add locally. Happy to contribute any of these here to help. FWIW I don't think we need another whole config file for this - just show the user to speficy `device=mps` in the readme +/ docs.

**Comment by bhupesh-sf** [comment]
Any update on this? Its being quite a time that support for MPS is underway but not concluding.

**Comment by SalmanMohammadi** [comment]
Closed due to #1706. For anyone following along, feel free to try out torchtune on MPS and let us know how you get along!
